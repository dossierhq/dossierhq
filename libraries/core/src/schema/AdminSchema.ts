import { ok, type ErrorType, type Result } from '../ErrorResult.js';
import { BaseSchema } from './BaseSchema.js';
import { PublishedSchema } from './PublishedSchema.js';
import type {
  AdminSchemaSpecification,
  AdminSchemaSpecificationUpdate,
  AdminSchemaSpecificationWithMigrations,
} from './SchemaSpecification.js';
import { schemaAdminToPublished } from './schemaAdminToPublished.js';
import { schemaUpdate } from './schemaUpdate.js';
import { schemaValidateAdmin } from './schemaValidateAdmin.js';

export class AdminSchema<
  TSpec extends
    | AdminSchemaSpecification
    | AdminSchemaSpecificationWithMigrations = AdminSchemaSpecification,
> extends BaseSchema<TSpec> {
  private cachedPublishedSchema: PublishedSchema | null = null;

  static createAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<AdminSchema, typeof ErrorType.BadRequest> {
    const result = AdminSchemaWithMigrations.createAndValidate(update);
    if (!result.isOk()) return result;

    const { migrations, ...specWithoutMigrations } = result.value.spec;

    return ok(new AdminSchema(specWithoutMigrations));
  }

  constructor(spec: TSpec) {
    super(spec);
  }

  validate(): Result<void, typeof ErrorType.BadRequest> {
    return schemaValidateAdmin(this);
  }

  toPublishedSchema(): PublishedSchema {
    if (this.cachedPublishedSchema) {
      return this.cachedPublishedSchema;
    }

    const publishedSpec = schemaAdminToPublished(this);

    this.cachedPublishedSchema = new PublishedSchema(publishedSpec);
    return this.cachedPublishedSchema;
  }
}

export class AdminSchemaWithMigrations extends AdminSchema<AdminSchemaSpecificationWithMigrations> {
  static override createAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<AdminSchemaWithMigrations, typeof ErrorType.BadRequest> {
    const emptySpec: AdminSchemaSpecificationWithMigrations = {
      schemaKind: 'admin',
      version: 0,
      entityTypes: [],
      valueTypes: [],
      patterns: [],
      indexes: [],
      migrations: [],
    };
    const empty = new AdminSchemaWithMigrations(emptySpec);
    return empty.updateAndValidate(update);
  }

  updateAndValidate(
    update: AdminSchemaSpecificationUpdate,
  ): Result<AdminSchemaWithMigrations, typeof ErrorType.BadRequest> {
    // Update
    const updatedResult = schemaUpdate(this.spec, update);
    if (updatedResult.isError()) return updatedResult;
    const updatedSpec = updatedResult.value;

    if (updatedSpec === this.spec) {
      return ok(this); // no change
    }

    // Validate
    const updatedSchema = new AdminSchemaWithMigrations(updatedSpec);
    const validateResult = updatedSchema.validate();
    if (validateResult.isError()) return validateResult;

    return ok(updatedSchema);
  }
}
