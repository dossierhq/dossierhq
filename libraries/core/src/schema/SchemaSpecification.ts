import type { LooseAutocomplete } from '../TypeUtils.js';
import type { EntityReference, Location, RichText, ValueItem } from '../Types.js';
import { RichTextNodeType } from '../Types.js';

export interface AdminEntityTypeSpecification {
  name: string;
  adminOnly: boolean;
  authKeyPattern: string | null;
  nameField: string | null;
  fields: AdminFieldSpecification[];
}

export interface AdminValueTypeSpecification {
  name: string;
  adminOnly: boolean;
  fields: AdminFieldSpecification[];
}

export interface AdminEntityTypeSpecificationUpdate {
  name: string;
  adminOnly?: boolean;
  authKeyPattern?: string | null;
  nameField?: string | null;
  fields: AdminFieldSpecificationUpdate[];
}

export interface AdminValueTypeSpecificationUpdate {
  name: string;
  adminOnly?: boolean;
  fields: AdminFieldSpecificationUpdate[];
}

export interface PublishedEntityTypeSpecification {
  name: string;
  authKeyPattern: string | null;
  fields: PublishedFieldSpecification[];
}

export interface PublishedValueTypeSpecification {
  name: string;
  fields: PublishedFieldSpecification[];
}

export const FieldType = {
  Boolean: 'Boolean',
  Entity: 'Entity',
  Location: 'Location',
  Number: 'Number',
  RichText: 'RichText',
  String: 'String',
  ValueItem: 'ValueItem',
} as const;
export type FieldType = (typeof FieldType)[keyof typeof FieldType];

interface SharedFieldSpecification {
  name: string;
  list: boolean;
  required: boolean;
}

export interface BooleanFieldSpecification extends SharedFieldSpecification {
  type: typeof FieldType.Boolean;
}

export interface EntityFieldSpecification extends SharedFieldSpecification {
  type: typeof FieldType.Entity;
  entityTypes: string[];
}

export interface LocationFieldSpecification extends SharedFieldSpecification {
  type: typeof FieldType.Location;
}

export interface NumberFieldSpecification extends SharedFieldSpecification {
  type: typeof FieldType.Number;
  integer: boolean;
}

export interface RichTextFieldSpecification extends SharedFieldSpecification {
  type: typeof FieldType.RichText;
  entityTypes: string[];
  linkEntityTypes: string[];
  valueTypes: string[];
  /** All node types are enabled if none are specified.
   *
   * The type can either be a standard RichTextNodeType or any type that's supported by the
   * application.
   */
  richTextNodes: LooseAutocomplete<RichTextNodeType>[];
}

export interface StringFieldSpecification extends SharedFieldSpecification {
  type: typeof FieldType.String;
  multiline: boolean;
  matchPattern: string | null;
  values: { value: string }[];
  index: string | null;
}

export interface ValueItemFieldSpecification extends SharedFieldSpecification {
  type: typeof FieldType.ValueItem;
  valueTypes: string[];
}

export type FieldSpecification =
  | BooleanFieldSpecification
  | EntityFieldSpecification
  | LocationFieldSpecification
  | NumberFieldSpecification
  | RichTextFieldSpecification
  | StringFieldSpecification
  | ValueItemFieldSpecification;

export type AdminFieldSpecification<TFieldSpec extends FieldSpecification = FieldSpecification> =
  TFieldSpec & {
    adminOnly: boolean;
  };

type PartialExcept<T, K extends keyof T> = Pick<T, K> & Partial<Omit<T, K>>;

export type AdminBooleanFieldSpecificationUpdate = PartialExcept<
  AdminFieldSpecification<BooleanFieldSpecification>,
  'name' | 'type'
>;
export type AdminEntityFieldSpecificationUpdate = PartialExcept<
  AdminFieldSpecification<EntityFieldSpecification>,
  'name' | 'type'
>;
export type AdminLocationFieldSpecificationUpdate = PartialExcept<
  AdminFieldSpecification<LocationFieldSpecification>,
  'name' | 'type'
>;
export type AdminNumberFieldSpecificationUpdate = PartialExcept<
  AdminFieldSpecification<NumberFieldSpecification>,
  'name' | 'type'
>;
export type AdminRichTextFieldSpecificationUpdate = PartialExcept<
  AdminFieldSpecification<RichTextFieldSpecification>,
  'name' | 'type'
>;
export type AdminStringFieldSpecificationUpdate = PartialExcept<
  AdminFieldSpecification<StringFieldSpecification>,
  'name' | 'type'
>;
export type AdminValueItemFieldSpecificationUpdate = PartialExcept<
  AdminFieldSpecification<ValueItemFieldSpecification>,
  'name' | 'type'
>;

export type AdminFieldSpecificationUpdate =
  | AdminBooleanFieldSpecificationUpdate
  | AdminEntityFieldSpecificationUpdate
  | AdminLocationFieldSpecificationUpdate
  | AdminNumberFieldSpecificationUpdate
  | AdminRichTextFieldSpecificationUpdate
  | AdminStringFieldSpecificationUpdate
  | AdminValueItemFieldSpecificationUpdate;

export type PublishedFieldSpecification = FieldSpecification;

export interface FieldValueTypeMap {
  [FieldType.Boolean]: boolean;
  [FieldType.Entity]: EntityReference;
  [FieldType.Location]: Location;
  [FieldType.Number]: number;
  [FieldType.RichText]: RichText;
  [FieldType.String]: string;
  [FieldType.ValueItem]: ValueItem;
}

export interface SchemaPatternSpecification {
  name: string;
  pattern: string;
}

export interface SchemaIndexSpecification {
  name: string;
  type: 'unique';
}

export interface PublishedSchemaSpecification {
  schemaKind: 'published';
  version: number;
  entityTypes: PublishedEntityTypeSpecification[];
  valueTypes: PublishedValueTypeSpecification[];
  patterns: SchemaPatternSpecification[];
  indexes: SchemaIndexSpecification[];
}

export interface AdminSchemaSpecification {
  schemaKind: 'admin';
  version: number;
  entityTypes: AdminEntityTypeSpecification[];
  valueTypes: AdminValueTypeSpecification[];
  patterns: SchemaPatternSpecification[];
  indexes: SchemaIndexSpecification[];
}

export interface AdminSchemaSpecificationWithMigrations extends AdminSchemaSpecification {
  migrations: AdminSchemaVersionMigration[];
}

export interface AdminSchemaVersionMigration {
  version: number;
  actions: AdminSchemaMigrationAction[];
}

export type AdminSchemaMigrationAction =
  | { action: 'renameType'; entityType: string; newName: string }
  | { action: 'renameType'; valueType: string; newName: string }
  | { action: 'renameField'; entityType: string; field: string; newName: string }
  | { action: 'renameField'; valueType: string; field: string; newName: string }
  | { action: 'deleteType'; entityType: string }
  | { action: 'deleteType'; valueType: string }
  | { action: 'deleteField'; entityType: string; field: string }
  | { action: 'deleteField'; valueType: string; field: string };

export type AdminSchemaTransientMigrationAction =
  | { action: 'renameIndex'; index: string; newName: string }
  | { action: 'deleteIndex'; index: string };

export interface AdminSchemaSpecificationUpdate {
  version?: number;
  entityTypes?: AdminEntityTypeSpecificationUpdate[];
  valueTypes?: AdminValueTypeSpecificationUpdate[];
  patterns?: SchemaPatternSpecification[];
  indexes?: SchemaIndexSpecification[];
  migrations?: AdminSchemaVersionMigration[];
  transientMigrations?: AdminSchemaTransientMigrationAction[];
}

export interface SchemaSpecificationUpdatePayload<
  TSpec extends
    | AdminSchemaSpecification
    | AdminSchemaSpecificationWithMigrations = AdminSchemaSpecification,
> {
  effect: 'updated' | 'none';
  schemaSpecification: TSpec;
}

export const REQUIRED_RICH_TEXT_NODES = [
  RichTextNodeType.root,
  RichTextNodeType.paragraph,
  RichTextNodeType.text,
  RichTextNodeType.linebreak,
  RichTextNodeType.tab,
];

export const GROUPED_RICH_TEXT_NODE_TYPES: RichTextNodeType[][] = [
  [RichTextNodeType.list, RichTextNodeType.listitem],
  [RichTextNodeType.code, RichTextNodeType['code-highlight']],
];
