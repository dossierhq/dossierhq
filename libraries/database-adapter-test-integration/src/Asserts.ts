import type { ErrorType, OkResult, Result } from '@jonasb/datadata-core';
import { createErrorResult, isFieldValueEqual } from '@jonasb/datadata-core';

class AssertionError extends Error {
  actual: unknown;
  expected: unknown;

  constructor(actual: unknown, expected: unknown, message: string) {
    super(message);
    this.name = 'AssertionError';
    this.actual = actual;
    this.expected = expected;
  }
}

export function assertTruthy(actual: unknown): asserts actual {
  if (!actual) {
    throw new AssertionError(actual, true, `Expected truthy`);
  }
}

export function assertSame<T>(actual: T, expected: T): void {
  if (actual !== expected) {
    throw new AssertionError(actual, expected, `Expected same, got ${actual} !== ${expected}`);
  }
}

export function assertNotSame<T>(actual: T, expected: T): void {
  if (actual === expected) {
    throw new AssertionError(actual, expected, `Expected not same, got ${actual} === ${expected}`);
  }
}

export function assertOkResult<TOk, TError extends ErrorType>(
  actual: Result<unknown, ErrorType>
): asserts actual is OkResult<TOk, TError> {
  if (actual.isError()) {
    throw new AssertionError(
      actual,
      undefined,
      `Expected ok, got error ${actual.error}: ${actual.message}`
    );
  }
}

export function assertErrorResult(
  actual: Result<unknown, ErrorType>,
  expectedErrorType: ErrorType,
  expectedMessage: string
): void {
  if (!actual.isError()) {
    throw new AssertionError(
      actual,
      createErrorResult(expectedErrorType, expectedMessage),
      `Expected error, but was ok`
    );
  }
  const expectedString = `${expectedErrorType}: ${expectedMessage}`;
  const actualString = `${actual.error}: ${actual.message}`;
  assertSame(actualString, expectedString);
}

export function assertResultValue<TOk, TError extends ErrorType>(
  result: Result<TOk, TError>,
  expectedValue: TOk
): void {
  assertOkResult(result);
  if (!isFieldValueEqual(result.value, expectedValue)) {
    throw new AssertionError(result.value, expectedValue, `Expected result values to equal`);
  }
}
