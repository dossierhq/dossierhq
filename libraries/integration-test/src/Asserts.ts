import type { ErrorResult, ErrorType, OkResult, Result } from '@dossierhq/core';
import { createErrorResult, isFieldValueEqual } from '@dossierhq/core';

class AssertionError extends Error {
  actual: unknown;
  expected: unknown;

  constructor(actual: unknown, expected: unknown, message: string) {
    super(message);
    this.name = 'AssertionError';
    this.actual = actual;
    this.expected = expected;
  }
}

export function assertTruthy(actual: unknown): asserts actual {
  if (!actual) {
    throw new AssertionError(actual, true, `Expected truthy`);
  }
}

export function assertSame<T>(actual: T, expected: T): void {
  if (actual !== expected) {
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    throw new AssertionError(actual, expected, `Expected same, got ${actual} !== ${expected}`);
  }
}

export function assertNotSame<T>(actual: T, expected: T): void {
  if (actual === expected) {
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    throw new AssertionError(actual, expected, `Expected not same, got ${actual} === ${expected}`);
  }
}

export function assertEquals<T>(actual: T, expected: T): void {
  //TODO need a different name for isFieldValueEqual
  if (!isFieldValueEqual(actual, expected)) {
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    throw new AssertionError(actual, expected, `Expected equals, got ${actual} !== ${expected}`);
  }
}

export function assertOkResult<TOk, TError extends ErrorType>(
  actual: Result<unknown, ErrorType>,
): asserts actual is OkResult<TOk, TError> {
  if (actual.isError()) {
    throw new AssertionError(
      actual,
      undefined,
      `Expected ok, got error ${actual.error}: ${actual.message}`,
    );
  }
}

export function assertErrorResult(
  actual: Result<unknown, ErrorType>,
  expectedErrorType: ErrorType,
  expectedMessage: string,
): asserts actual is ErrorResult<unknown, ErrorType> {
  if (!actual.isError()) {
    throw new AssertionError(
      actual,
      createErrorResult(expectedErrorType, expectedMessage),
      `Expected error, but was ok`,
    );
  }
  const expectedString = `${expectedErrorType}: ${expectedMessage}`;
  const actualString = `${actual.error}: ${actual.message}`;
  assertSame(actualString, expectedString);
}

export function assertResultValue<TOk, TError extends ErrorType>(
  result: Result<TOk, TError>,
  expectedValue: TOk,
): asserts result is OkResult<TOk, TError> {
  assertOkResult(result);
  if (!isFieldValueEqual(result.value, expectedValue)) {
    throw new AssertionError(result.value, expectedValue, `Expected result values to equal`);
  }
}

export function assertExhaustive(param: never): never {
  // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
  throw new Error(`Invalid exhaustiveness check: ${param}`);
}
