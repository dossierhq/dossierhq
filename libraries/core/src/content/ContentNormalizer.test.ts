import { describe, expect, test } from 'vitest';
import { AdminSchema } from '../schema/AdminSchema.js';
import { FieldType } from '../schema/SchemaSpecification.js';
import { assertIsDefined } from '../utils/Asserts.js';
import {
  normalizeEntityFields,
  normalizeFieldValue,
  normalizeValueItem,
} from './ContentNormalizer.js';
import {
  createRichText,
  createRichTextParagraphNode,
  createRichTextValueItemNode,
} from './RichTextUtils.js';

const schema = AdminSchema.createAndValidate({
  entityTypes: [
    {
      name: 'Foo',
      fields: [
        { name: 'string', type: FieldType.String },
        { name: 'stringList', type: FieldType.String, list: true },
        { name: 'twoStrings', type: FieldType.ValueItem, valueTypes: ['TwoStrings'] },
        { name: 'richText', type: FieldType.RichText },
      ],
    },
  ],
  valueTypes: [
    {
      name: 'TwoStrings',
      fields: [
        { name: 'string1', type: FieldType.String },
        { name: 'string2', type: FieldType.String },
      ],
    },
  ],
}).valueOrThrow();

function getEntityFieldSpec(schema: AdminSchema, entityType: string, fieldName: string) {
  const entitySpec = schema.getEntityTypeSpecification(entityType);
  assertIsDefined(entitySpec);
  const fieldSpec = schema.getEntityFieldSpecification(entitySpec, fieldName);
  assertIsDefined(fieldSpec);
  return fieldSpec;
}

describe('normalizeEntityFields', () => {
  test('empty Foo', () => {
    expect(
      normalizeEntityFields(schema, ['entity'], {
        info: { type: 'Foo' },
        fields: {},
      }).valueOrThrow(),
    ).toMatchSnapshot();
  });

  test('empty Foo (excludeOmittedEntityFields)', () => {
    expect(
      normalizeEntityFields(
        schema,
        ['entity'],
        { info: { type: 'Foo' }, fields: {} },
        { excludeOmittedEntityFields: true },
      ).valueOrThrow(),
    ).toEqual({});
  });

  test('empty fields in value item', () => {
    expect(
      normalizeEntityFields(schema, ['entity'], {
        info: { type: 'Foo' },
        fields: { twoStrings: { type: 'TwoStrings', string1: '', string2: null } },
      }).valueOrThrow(),
    ).toMatchSnapshot();
  });

  test('empty fields in value item in rich text', () => {
    expect(
      normalizeEntityFields(schema, ['entity'], {
        info: { type: 'Foo' },
        fields: {
          richText: createRichText([
            createRichTextValueItemNode({ type: 'TwoStrings', string1: '', string2: null }),
          ]),
        },
      }).valueOrThrow(),
    ).toMatchSnapshot();
  });
});

describe('normalizeValueItem', () => {
  test('empty TwoStrings', () => {
    expect(
      normalizeValueItem(schema, ['entity'], { type: 'TwoStrings' }).valueOrThrow(),
    ).toMatchSnapshot();
  });

  test('TwoStrings with empty strings', () => {
    expect(
      normalizeValueItem(schema, ['entity'], {
        type: 'TwoStrings',
        string1: '',
        string2: null,
      }).valueOrThrow(),
    ).toMatchSnapshot();
  });
});

describe('normalizeFieldValue()', () => {
  test('"" => null', () => {
    expect(normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'string'), '')).toEqual(
      null,
    );
  });

  test('[] => null', () => {
    expect(normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), [])).toBe(
      null,
    );
  });

  test('[string, ""] => [string]', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), ['hello', '']),
    ).toEqual(['hello']);
  });

  test('[string] => [string] (no change)', () => {
    const fieldValue = ['hello', 'world'];
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), fieldValue),
    ).toBe(fieldValue);
  });

  test('{string1:string,string2:""} => {string1:string,string2:null}', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), {
        type: 'TwoStrings',
        string1: 'Hello',
        string2: '',
      }),
    ).toEqual({ type: 'TwoStrings', string1: 'Hello', string2: null });
  });

  test('{string1:undefined} => {string1:null,string2:null}', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), {
        type: 'TwoStrings',
        string1: undefined,
        // no string2
      }),
    ).toEqual({ type: 'TwoStrings', string1: null, string2: null });
  });

  test('{string1:string,string2:string} => {string1:string,string2:string} (no change)', () => {
    const fieldValue = {
      type: 'TwoStrings',
      string1: 'Hello',
      string2: 'World',
    };
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), fieldValue),
    ).toBe(fieldValue);
  });

  test('string undefined => null', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'string'), undefined),
    ).toBe(null);
  });

  test('string[] undefined => null', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'stringList'), undefined),
    ).toBe(null);
  });

  test('ValueItem: undefined => null', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'twoStrings'), undefined),
    ).toBe(null);
  });

  test('RichText: empty paragraph => null', () => {
    expect(
      normalizeFieldValue(
        schema,
        getEntityFieldSpec(schema, 'Foo', 'richText'),
        createRichText([createRichTextParagraphNode([])]),
      ),
    ).toBe(null);
  });

  test('RichText: let invalid rich text pass through (no root)', () => {
    expect(normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'richText'), {})).toEqual(
      {},
    );
  });

  test('RichText: let invalid rich text pass through (string in root)', () => {
    expect(
      normalizeFieldValue(schema, getEntityFieldSpec(schema, 'Foo', 'richText'), {
        root: 'hello world',
      }),
    ).toEqual({ root: 'hello world' });
  });
});
