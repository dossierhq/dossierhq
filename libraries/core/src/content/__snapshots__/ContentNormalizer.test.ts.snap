// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`normalizeEntityFields > empty Foo 1`] = `
{
  "richText": null,
  "string": null,
  "stringList": null,
  "twoStrings": null,
}
`;

exports[`normalizeEntityFields > empty fields in value item 1`] = `
{
  "richText": null,
  "string": null,
  "stringList": null,
  "twoStrings": {
    "string1": null,
    "string2": null,
    "type": "TwoStrings",
  },
}
`;

exports[`normalizeEntityFields > empty fields in value item in rich text 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "data": {
            "string1": null,
            "string2": null,
            "type": "TwoStrings",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
  "string": null,
  "stringList": null,
  "twoStrings": null,
}
`;

exports[`normalizeValueItem > TwoStrings with empty strings 1`] = `
{
  "string1": null,
  "string2": null,
  "type": "TwoStrings",
}
`;

exports[`normalizeValueItem > empty TwoStrings 1`] = `
{
  "string1": null,
  "string2": null,
  "type": "TwoStrings",
}
`;
